lms/services/youtube.py

```diff
@@ -1,21 +1,74 @@
+import json
+
+from lms.services import SerializableError
+from lms.services.http import HTTPService
+
+
+class VideoNotFound(SerializableError):
+    def __init__(self, video_id):
+        super().__init__(error_code="youtube_video_not_found", details={"video_id": video_id})
+
+
 class YoutubeService:
-    """An interface for dealing with Youtube API."""
+    """An interface for dealing with YouTube API."""
 
-    def __init__(self, enabled: bool):
+    def __init__(self, enabled: bool, api_url: str, api_key: str):
         """
-        Initialise the Youtube service.
+        Initialise the YouTube service.
 
-        :param enabled: Whether Youtube is enabled on this instance
+        :param enabled: Whether YouTube is enabled on this instance
+        :param api_url: YouTube API base URL
+        :param api_key: YouTube API key
         """
         self._enabled = enabled
+        self._api_url = api_url
+        self._api_key = api_key
+        self._http = HTTPService()
 
     @property
     def enabled(self) -> bool:
-        """Get whether this instance is configured for Youtube."""
+        """Get whether this instance is configured for YouTube."""
+
+        return bool(self._enabled and self._api_url and self._api_url)
+
+    def video_info(self, video_id: str) -> dict:
+        """
+        Fetch YouTube video information.
+
+        :param video_id: A YouTube video ID
+        :raise VideoNotFound: If the video cannot be found
+        """
+
+        resp = self._http.get(url=f"{self._api_url}/videos", params={
+            "id": video_id,
+            "key": self._api_key,
+            "part": "contentDetails,snippet",
+            "maxResults": "1",
+        })
 
-        return self._enabled
+        # TODO Be more strict on response schema
+        json_resp: dict = json.loads(resp.text)
 
+        try:
+            item = json_resp["items"][0]
+        except IndexError:
+            raise VideoNotFound(video_id)
 
-def factory(_context, request):
+        snippet = item["snippet"]
+        return {
+            "image": snippet["thumbnails"]["medium"]["url"],
+            "title": snippet["title"],
+            "channel": snippet["channelTitle"],
+            "duration": item["contentDetails"]["duration"]  # ISO duration
+        }
+
+
+def factory(_context, request) -> YoutubeService:
     ai_settings = request.lti_user.application_instance.settings
-    return YoutubeService(enabled=ai_settings.get("youtube", "enabled"))
+    app_settings = request.registry.settings
+
+    return YoutubeService(
+        enabled=ai_settings.get("youtube", "enabled"),
+        api_url=app_settings.get("youtube_api_url"),
+        api_key=app_settings.get("youtube_api_key"),
```

Comment:
I'd do something like 

```http_service=request.find_service(name="http")``` here instead of building it `__init__` the jstor one seems to be doing that but I'd say it's an outlier.

Reply:
I did it like this because I saw it in JSTOR and thought it was intentional, but I do agree it's better to explicitly inject dependencies.